pkg/loki/modules.go

```diff
@@ -654,15 +654,12 @@ func (t *Loki) initStore() (_ services.Service, err error) {
 			t.Cfg.Ingester.QueryStoreMaxLookBackPeriod = mlb
 		}
 
-		if asyncStore {
-			t.Cfg.StorageConfig.EnableAsyncStore = true
-			t.Cfg.StorageConfig.AsyncStoreConfig = storage.AsyncStoreCfg{
-				IngesterQuerier: t.ingesterQuerier,
-				QueryIngestersWithin: calculateAsyncStoreQueryIngestersWithin(
-					t.Cfg.Querier.QueryIngestersWithin,
-					minIngesterQueryStoreDuration,
-				),
-			}
+		t.Cfg.StorageConfig.AsyncStoreConfig = storage.AsyncStoreCfg{
+			IngesterQuerier: t.ingesterQuerier,
```

Comment:
wondering how this assignment works given that the store has no dependency on ingesterQuerier. It would be possible for initStore to get called before ingesterQuerier is initialised

I am sure i must be missing something given that this code already works

Reply:
This is a great question. It looks like it's been added in a couple of places.  

It works because it's a dependency of [`Querier`](https://github.com/grafana/loki/blob/358b8965c4fe5e6d95d09aa3166a02213974bd64/pkg/loki/loki.go#L655) and [`RuleEvaluator`](https://github.com/grafana/loki/blob/358b8965c4fe5e6d95d09aa3166a02213974bd64/pkg/loki/loki.go#L660) which are both dependencies of `All`, `Read`, `Ruler`, and `Backend`

It's also added as a dependency [here](https://github.com/grafana/loki/blob/358b8965c4fe5e6d95d09aa3166a02213974bd64/pkg/loki/loki.go#L706-L709). I think this is redundant as it's in the above graph, now.

Reply:
After looking at it some more, I just made `IngesterQuerier` a dependency of `Store`. Before, they were peers so `IngesterQuerier` might not exist when `initStore` was called. Now the dependency is explicit.